AJ-1283: Add `defaultGoogleBucketOptions` and `defaultAzureStorageOptions` to `testData`

jladieu opened the pull request:
### Jira Ticket: https://broadworkbench.atlassian.net/browse/AJ-1283

<!-- ### Dependencies --->
This is a prefactor to improve reuse of exports I'll need while introducing test coverage to safeguard a changes to `Data.js`.  It is self-contained refactoring in test-only code & has no dependencies on other work.

## Summary of changes:
Extract default workspace test data into shared testing lib

Since these defaults are useful beyond just analysis, bringing them up a level makes them reusable to the whole codebase without requiring dependencies into the `analysis` subdir.

### Testing strategy
- [x] Existing unit tests that use these constructs continue to pass.  Production codepaths are untouched.

<!-- ### Visual Aids -->
N/A

jladieu committed:
Extract default workspace test data into shared testing lib

Since these defaults are useful beyond just analysis, bringing them up a level makes them accessible to the whole codebase.

jdcanas reviewed (changes requested):
This does not seem desirable, note that the `testData.ts` file lives within the directory for analysis components.

I think we should be careful with test data that spans functional areas of the code base, and keep test data files small by scoping them as much as possible. I can see the use case for these though, and therefore I would recommend a different test data file at a higher level than the analysis folder for this.

nawatts commented on src/analysis/_testData/testData.ts:
```diff
@@ -215,6 +216,18 @@ export const defaultTestDisk = {
   blockSize: 4096,
 };
 
+export const defaultGoogleBucketOptions = {
```
```suggestion
export const defaultGoogleBucketOptions: SomeType = {
```

It would be nice to add types to these to these to ensure that the test data stays up to date with the actual interface.

jladieu commented:
> This does not seem desirable, note that the `testData.ts` file lives within the directory for analysis components.
> 
> I think we should be careful with test data that spans functional areas of the code base, and keep test data files small by scoping them as much as possible. I can see the use case for these though, and therefore I would recommend a different test data file at a higher level than the analysis folder for this.

Ahh I didn't realize I was crossing over a bad boundary with this extraction.  I figured the `testData` was fair use across a broader area.

I'll hold off.

nawatts commented:
Agreed with @jdcanas about moving general workspace related fixtures to a higher level. Ideally, I think that includes the `defaultGoogleWorkspace`, `defaultAzureWorkspace`, etc. that are currently in `analysis/_testData/testData.ts`.

jladieu commented:
Oops, sorry for the premature close here -- I misunderstood @jdcanas' feedback.

I think picking a location boils down to a naming/organization problem, and I'm definitely not super familiar with the conventions here...might `src/testing/common-defaults.ts` be the right spot & name?  Happy to get better suggestions.

nawatts commented:
I think my vote would be for `src/testing/workspace-fixtures.ts` or something like that.

nawatts commented on src/analysis/_testData/testData.ts:
```diff
@@ -120,23 +118,6 @@ export const imageDocs = [
   },
 ];
 
-export const defaultGoogleWorkspace: GoogleWorkspace = {
```
It looks like the moved objects will need to imported into this module since they're used by other things here.

jladieu commented:
@jdcanas thanks for the suggestion -- I pulled the changes up into `src/testing/workspace-fixtures.ts`

Small bookkeeping note for @nawatts: I'm doing this refactoring to reduce resistance to this [ticket](https://broadworkbench.atlassian.net/browse/AJ-1283?atlOrigin=eyJpIjoiMjdkOGNhNjBlMTYwNGM4YWEwZWUyZGJkNzE0ZGNkYTQiLCJwIjoiaiJ9); should I just tag this PR as part of that ticket, or, since it's more generic and essentially optional, would this belong under UIE?

nawatts commented on src/testing/workspace-fixtures.ts:
```diff
@@ -0,0 +1,52 @@
+  canCompute: true,
+};
+
+export const defaultAzureStorageOptions = {
```
If this is to be the "blessed" default fixture for this type of data, I'd prefer it have some reasonably realistic values instead of everything being undefined.

Similar to what's in the ContextBar test:
```ts
const defaultAzureStorageOptions = {
  azureContainerRegion: 'eastus',
  azureContainerUrl: 'container-url',
  azureContainerSasUrl: 'container-url?sas',
};
```

jladieu replied:
Happy to put something legit in here, but I actively avoided reusing the one in `ContextBar` because it's the only one that did so.  Most of the other tests are focused on GCP and therefore leaving `azureStorageOptions` undefined.

One thought: to make this explicit, I could provide another export called `emptyAzureStorageOptions`, and then refer to that throughout the tests that are focused on GCP.  But before I take that on I want to make sure it's reasonable.  WDYT?

nawatts replied:
> Most of the other tests are focused on GCP and therefore leaving azureStorageOptions undefined.

I think this will become less true over time as we add more Azure functionality.

jladieu replied:
Done.  Most of the tests were happy to use the more intelligent defaults.  A couple tests explicitly wanted `emptyAzureStorageOptions`.

cahrens replied:
Note that our code should handle the case where these values are not yet known. Storage-related values depend on an endpoint that initially returns empty values during workspace cloning, for example.

The IA code has historically shown default values for Google workspaces until the actual values can be fetched from Google… and in the case of request pays workspaces, the real values can't be fetched at all. I instead think we need to be doing the more correct approach of handling the "unknown" case and displaying something sensible to the user when we either haven't heard back or can't fetch the actual the values.

cahrens replied:
Looking more at this PR, I think there is a misunderstanding about what I meant by "default" in Dashboard.test.js. I meant that these are the default values the code expects/will use until the initial server response is received.

jladieu replied:
I agree with @cahrens pushback on this and I've switched the shared `defaultAzureStorageOptions` over to be `undefined` to closely track the actual implemented defaults, and included a comment explaining why.  This means I backed up from @nawatts suggestion to promote the ContextBar test data to this fixture, and instead moved it back inline with a name that doesn't conflict with and get confused for being the _default_ behavior see [comment](https://github.com/DataBiosphere/terra-ui/pull/4157/files#r1311747337).

This also means `emptyAzureStorageOptions` gets nixed.

jladieu commented on src/testing/workspace-fixtures.ts:
```diff
@@ -0,0 +1,58 @@
+  azureContainerSasUrl: 'container-url?sas',
+};
+
+export const emptyAzureStorageOptions = {
```
This is provided for a couple test suites which are explicitly checking for scenarios where only the GCP attributes are initialized.

jladieu commented:
@jdcanas thanks for the earlier feedback -- please take a look at the result; a couple things happened since the initial push:

1) I bumped the helpers up into `testing/workspace-fixtures.ts`, which required updating the imports from the old location throughout
2) I added smarter defaults to `defaultAzureStorageOptions` to reflect a valid state
3) I added another export `emptyAzureStorageOptions` to cover some test scenarios that were explicitly checking GCP-only configurations

cahrens commented on src/testing/workspace-fixtures.ts:
```diff
@@ -0,0 +1,58 @@
+};
+
+export const defaultAzureStorageOptions = {
+  azureContainerRegion: 'eastus',
```
Note that `eastus` will not be the default soon. Do we need this to be a specific value, or can it be a dummy value?

cahrens commented on src/pages/workspaces/workspace/Dashboard.test.js:
```diff
@@ -100,17 +99,6 @@ describe('WorkspaceNotifications', () => {
-  azureContainerSasUrl: undefined,
-};
-
 describe('BucketLocation', () => {
```
All of these tests in `BucketLocation` are Google-specific. Thus all of the Azure-related storage values should ALWAYS be undefined… so it is not correct to replace them with the new `defaultAzureStorageOptions`.

I would prefer to the have the workspaces tests restored back to how they were before. Those "default" values are meant to represent the initial state of a workspace and/or the values for cloud-specific storage when the workspace being rendered is of the "other" type. If we are rendering a GCP workspace, for example, the Azure values should all remain as undefined.

cahrens replied:
I defined these as variables just to make my test code easier to read/more concise, but I didn't intend them to become shared across multiple tests.

jladieu replied:
As discussed in Slack, I'm happy to re-inline these if you want and then this file doesn't get changed at all. But for the time being at least your definition of them is perfect for what I needed, and does a good job of tracking the actual behavior defined in `useWorkspace`, so I'd like to keep them available in workspace-fixtures.ts if possible.

jladieu commented on src/analysis/ContextBar.test.ts:
```diff
@@ -379,12 +377,7 @@ const runtimeDisk: DecoratedPersistentDisk = {
-  fetchedGoogleBucketLocation: undefined,
-};
-const defaultAzureStorageOptions = {
+const populatedAzureStorageOptions = {
```
I initially extracted `defaultAzureStorageOptions` as well up into the shared workspace-fixtures module I added for this, but after some feedback from @cahrens concluded that calling these defaults in that shared domain isn't accurate.  The defaults are actually all empty.

So I updated `defaultAzureStorageOptions` to clarify that the fields are unset, and renamed this to disambiguate & distinguish it from the defaults.

nawatts reviewed (approved):
I suspect the original "default" in the names of these variables meant "default for IA tests" and doesn't carry over as well to their expanded scope.

cahrens reviewed (approved):
Thanks for making the changes! In the future, please don't squash commits once a review is in progress because it makes it more difficult for reviewers to see what has changed since they last looked at the PR. And we automatically squash commits on merge, so you don't need to worry about the commit history including minor updates.

jladieu committed:
Update a missed import.
